chain/network/src/routing/actor.rs

```diff
@@ -0,0 +1,319 @@
+use crate::private_actix::{StopMsg, ValidateEdgeList};
+use crate::routing;
+use crate::routing::edge_validator_actor::EdgeValidatorActor;
+use crate::stats::metrics;
+use crate::store;
+use actix::{
+    ActorContext as _, ActorFutureExt, Addr, Context, ContextFutureSpawner as _, Running,
+    WrapFuture as _,
+};
+use near_network_primitives::time;
+use near_network_primitives::types::Edge;
+use near_performance_metrics_macros::perf;
+use near_primitives::network::PeerId;
+use near_rate_limiter::{ActixMessageResponse, ActixMessageWrapper, ThrottleToken};
+use parking_lot::RwLock;
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+use tracing::warn;
+
+/// Actor that maintains routing table information.
+///
+/// We store the following information
+///   - routing information (where a message should be send to reach given peer)
+///
+/// We use store for following reasons:
+///   - store removed edges to disk
+///   - we currently don't store active edges to disk
+pub(crate) struct Actor {
+    clock: time::Clock,
+    my_peer_id: PeerId,
+    // Graph is locked for writes only by this Actor, and only to update a bunch of edges.
+    // You must not put expensive computations under the lock and/or DB access.
+    // TODO: Reimplement GraphWithCache to GraphWithCache(Arc<RwLock<GraphWithCacheInner>>),
+    // to enforce the constraint above.
+    graph: Arc<RwLock<routing::GraphWithCache>>,
+    store: store::Store,
+    /// Last time a peer was reachable.
+    peer_reachable_at: HashMap<PeerId, time::Instant>,
+    /// List of Peers to ban
+    peers_to_ban: Vec<PeerId>,
+    /// EdgeValidatorActor, which is responsible for validating edges.
+    edge_validator_pool: Addr<EdgeValidatorActor>,
+    /// Number of edge validations in progress; We will not update routing table as long as
+    /// this number is non zero.
+    edge_validator_requests_in_progress: u64,
+}
+
+impl Actor {
+    pub fn new(
+        clock: time::Clock,
+        store: store::Store,
+        graph: Arc<RwLock<routing::GraphWithCache>>,
+    ) -> Self {
+        let my_peer_id = graph.read().my_peer_id();
+        Self {
+            clock,
+            my_peer_id,
+            graph,
+            store,
+            peer_reachable_at: Default::default(),
+            peers_to_ban: Default::default(),
+            edge_validator_requests_in_progress: 0,
+            edge_validator_pool: actix::SyncArbiter::start(4, || EdgeValidatorActor {}),
+        }
+    }
+
+    /// Add several edges to the current view of the network.
+    /// These edges are assumed to have been verified at this point.
+    /// Return list of edges added. Requires DB access.
```

Comment:
so we always return the same list of that we received ? or only "new" ones ?

Reply:
each "edge" is actually a version of an edge, identified by nonce.
We return edge versions not previously observed, so that PeerManager can broadcast only stuff that we didn't see yet. I've updated the comment.